File: aws/src/handlers/rooms_handler.py
import json
import boto3
import os
import base64
from cors_utils import cors_response, cors_error_response, handle_preflight_request, extract_origin_from_event
import uuid
from datetime import datetime, timedelta
from decimal import Decimal
from botocore.exceptions import ClientError

# Initialize AWS clients
dynamodb = boto3.resource('dynamodb')
s3 = boto3.client('s3')

# Environment variables
ROOMS_TABLE = os.environ['DYNAMODB_TABLE_ROOMS']
BOOKINGS_TABLE = os.environ['DYNAMODB_TABLE_BOOKINGS']
S3_BUCKET = os.environ.get('S3_BUCKET', 'dogotel-images')

rooms_table = dynamodb.Table(ROOMS_TABLE)
bookings_table = dynamodb.Table(BOOKINGS_TABLE)

def lambda_handler(event, context):
    """Main Lambda handler"""
    try:
        http_method = event['httpMethod']
        path = event['path']
        origin = extract_origin_from_event(event)
        
        print(f"Processing {http_method} {path} from origin: {origin}")
        
        # Handle preflight OPTIONS requests
        if http_method == 'OPTIONS':
            return handle_preflight_request(event, origin=origin)
        
        print(f"Processing {http_method} {path}")
        
        if path == '/rooms' and http_method == 'GET':
            return handle_list_rooms(event)
        elif path.startswith('/rooms/') and path.count('/') == 2 and http_method == 'GET':
            room_id = path.split('/')[-1]
            return handle_get_room(room_id, event)
        elif path.startswith('/rooms/') and path.endswith('/unavailable-dates') and http_method == 'GET':
            room_id = path.split('/')[-2]
            return handle_get_unavailable_dates(room_id, event)
        elif path.startswith('/rooms/') and path.endswith('/unavailable-ranges') and http_method == 'GET':
            room_id = path.split('/')[-2]
            return handle_get_unavailable_ranges(room_id, event)
        elif path == '/rooms' and http_method == 'POST':
            return handle_create_room(event)
        elif path.startswith('/rooms/') and path.count('/') == 2 and http_method == 'PUT':
            room_id = path.split('/')[-1]
            return handle_update_room(room_id, event)
        elif path.startswith('/rooms/') and path.count('/') == 2 and http_method == 'DELETE':
            room_id = path.split('/')[-1]
            return handle_delete_room(room_id, event)
        else:
            return cors_error_response(404, 'Endpoint not found', origin)
            
    except Exception as e:
        print(f"Error in rooms handler: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_list_rooms(event):
    """Handle listing all rooms with optional filters"""
    try:
        origin = extract_origin_from_event(event)
        query_params = event.get('queryStringParameters') or {}
        min_price = query_params.get('min_price')
        max_price = query_params.get('max_price')
        check_in = query_params.get('check_in')
        check_out = query_params.get('check_out')
        size = query_params.get('size')
        
        # Get all rooms
        response = rooms_table.scan()
        rooms = response['Items']
        
        # Apply filters
        filtered_rooms = []
        for room in rooms:
            # Price filter (using price field to match frontend)
            if min_price and room.get('price', room.get('price_per_night', 0)) < Decimal(min_price):
                continue
            if max_price and room.get('price', room.get('price_per_night', 0)) > Decimal(max_price):
                continue
            
            # Size filter
            if size and room.get('size', '').lower() != size.lower():
                continue
            
            # Availability filter
            if check_in and check_out:
                if not is_room_available(room['room_id'], check_in, check_out):
                    continue
            
            # Convert Decimal to float for JSON serialization and transform to frontend format
            room_copy = transform_room_for_frontend(convert_decimals(room))
            filtered_rooms.append(room_copy)
        
        # Return direct array to match frontend expectations
        return cors_response(200, filtered_rooms, origin)
        
    except Exception as e:
        print(f"List rooms error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_get_room(room_id, event):
    """Handle getting a specific room"""
    try:
        origin = extract_origin_from_event(event)
        
        response = rooms_table.get_item(Key={'room_id': room_id})
        
        if 'Item' not in response:
            return cors_error_response(404, 'Room not found', origin)
        
        room = transform_room_for_frontend(convert_decimals(response['Item']))
        
        return cors_response(200, room, origin)
        
    except Exception as e:
        print(f"Get room error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_get_unavailable_dates(room_id, event):
    """Handle getting unavailable dates for a specific room"""
    try:
        origin = extract_origin_from_event(event)
        
        # Get all confirmed bookings for this room
        response = bookings_table.query(
            IndexName='RoomBookingsIndex',
            KeyConditionExpression='room_id = :room_id',
            FilterExpression='#status IN (:confirmed, :checked_in)',
            ExpressionAttributeNames={'#status': 'status'},
            ExpressionAttributeValues={
                ':room_id': room_id,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in'
            }
        )
        
        unavailable_dates = []
        for booking in response['Items']:
            # Generate all dates between check_in and check_out
            check_in = datetime.fromisoformat(booking['check_in'].replace('Z', '+00:00'))
            check_out = datetime.fromisoformat(booking['check_out'].replace('Z', '+00:00'))
            
            current_date = check_in.date()
            end_date = check_out.date()
            
            while current_date < end_date:  # Don't include checkout date
                unavailable_dates.append(current_date.isoformat())
                current_date += timedelta(days=1)
        
        return cors_response(200, {
            'roomId': room_id,
            'unavailableDates': sorted(list(set(unavailable_dates)))
        }, origin)
        
    except Exception as e:
        print(f"Get unavailable dates error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_get_unavailable_ranges(room_id, event):
    """Handle getting unavailable date ranges for a specific room"""
    try:
        origin = extract_origin_from_event(event)
        
        # Get all confirmed bookings for this room
        response = bookings_table.query(
            IndexName='RoomBookingsIndex',
            KeyConditionExpression='room_id = :room_id',
            FilterExpression='#status IN (:confirmed, :checked_in)',
            ExpressionAttributeNames={'#status': 'status'},
            ExpressionAttributeValues={
                ':room_id': room_id,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in'
            }
        )
        
        unavailable_ranges = []
        for booking in response['Items']:
            check_in = booking['check_in']
            check_out = booking['check_out']
            
            # Convert to date format for ranges
            check_in_date = datetime.fromisoformat(check_in.replace('Z', '+00:00')).date()
            check_out_date = datetime.fromisoformat(check_out.replace('Z', '+00:00')).date()
            
            # Adjust checkout date (booking ends day before checkout)
            actual_end_date = check_out_date - timedelta(days=1)
            
            unavailable_ranges.append({
                'start': check_in_date.isoformat(),
                'end': actual_end_date.isoformat()
            })
        
        return cors_response(200, {
            'roomId': room_id,
            'unavailableRanges': unavailable_ranges
        }, origin)
        
    except Exception as e:
        print(f"Get unavailable ranges error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_create_room(event):
    """Handle creating a new room - exact format from Ran_dev.txt"""
    try:
        origin = extract_origin_from_event(event)
        
        body = json.loads(event['body'])
        room_id = str(uuid.uuid4())
        
        # Required fields from Ran_dev.txt
        required_fields = ['title', 'description', 'dogsAmount', 'price', 'size']
        for field in required_fields:
            if field not in body:
                return cors_error_response(400, f'{field} is required', origin)
        
        # Handle image upload to S3 if it's base64
        image_url = body.get('image', '')
        if image_url and image_url.startswith('data:image/'):
            image_url = upload_image_to_s3(image_url, f"rooms/{room_id}")
        
        room_data = {
            'room_id': room_id,
            'title': body['title'],
            'subtitle': body.get('subtitle', ''),
            'description': body['description'],
            'dogsAmount': int(body['dogsAmount']),
            'price': int(body['price']),
            'size': body['size'],
            'image': image_url,
            'reviews': body.get('reviews', []),
            'features': body.get('features', []),
            'included': body.get('included', []),
            'createdAt': datetime.utcnow().isoformat(),
            'updatedAt': datetime.utcnow().isoformat()
        }
        
        rooms_table.put_item(Item=room_data)
        
        return cors_response(201, {
                'success': True,
                'id': room_id
            }, origin)
        
    except Exception as e:
        print(f"Create room error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_update_room(room_id, event):
    """Handle updating an existing room - exact format from Ran_dev.txt"""
    try:
        origin = extract_origin_from_event(event)
        
        body = json.loads(event['body'])
        
        # Get current room
        response = rooms_table.get_item(Key={'room_id': room_id})
        if 'Item' not in response:
            return cors_error_response(404, 'Room not found', origin)
        
        # Handle image upload to S3 if it's base64
        image_url = body.get('image', '')
        if image_url and image_url.startswith('data:image/'):
            image_url = upload_image_to_s3(image_url, f"rooms/{room_id}")
        
        # Update fields if provided
        update_expression = []
        expression_attribute_names = {}
        expression_attribute_values = {}
        
        updatable_fields = {
            'title': 'title',
            'subtitle': 'subtitle', 
            'description': 'description',
            'dogsAmount': 'dogsAmount',
            'price': 'price',
            'size': 'size',
            'image': 'image',
            'features': 'features',
            'included': 'included'
        }
        
        for field, db_field in updatable_fields.items():
            if field in body:
                update_expression.append(f"#{db_field} = :{db_field}")
                expression_attribute_names[f"#{db_field}"] = db_field
                if field == 'dogsAmount':
                    expression_attribute_values[f":{db_field}"] = int(body[field])
                elif field == 'price':
                    expression_attribute_values[f":{db_field}"] = int(body[field])
                elif field == 'image' and image_url:
                    expression_attribute_values[f":{db_field}"] = image_url
                else:
                    expression_attribute_values[f":{db_field}"] = body[field]
        
        # Always update the updatedAt timestamp
        update_expression.append("#updatedAt = :updatedAt")
        expression_attribute_names["#updatedAt"] = "updatedAt"
        expression_attribute_values[":updatedAt"] = datetime.utcnow().isoformat()
        
        rooms_table.update_item(
            Key={'room_id': room_id},
            UpdateExpression="SET " + ", ".join(update_expression),
            ExpressionAttributeNames=expression_attribute_names,
            ExpressionAttributeValues=expression_attribute_values
        )
        
        return cors_response(200, {
                'success': True
            }, origin)
        
    except Exception as e:
        print(f"Update room error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_delete_room(room_id, event):
    """Handle deleting a room - from Ran_dev.txt"""
    try:
        origin = extract_origin_from_event(event)
        
        # Check if room exists
        response = rooms_table.get_item(Key={'room_id': room_id})
        if 'Item' not in response:
            return cors_error_response(404, 'Room not found', origin)
        
        # Check for active bookings (optional safety check)
        booking_response = bookings_table.query(
            IndexName='RoomBookingsIndex',
            KeyConditionExpression='room_id = :room_id',
            FilterExpression='#status IN (:confirmed, :checked_in)',
            ExpressionAttributeNames={'#status': 'status'},
            ExpressionAttributeValues={
                ':room_id': room_id,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in'
            }
        )
        
        if booking_response['Items']:
            return cors_error_response(400, 'Cannot delete room with active bookings', origin)
        
        # Delete room
        rooms_table.delete_item(Key={'room_id': room_id})
        
        return cors_response(200, {'success': True}, origin)
        
    except Exception as e:
        print(f"Delete room error: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def is_room_available(room_id, check_in, check_out):
    """Check if room is available for given dates"""
    try:
        response = bookings_table.query(
            IndexName='RoomBookingsIndex',
            KeyConditionExpression='room_id = :room_id',
            FilterExpression='#status IN (:confirmed, :checked_in) AND (check_in <= :check_out AND check_out >= :check_in)',
            ExpressionAttributeNames={'#status': 'status'},
            ExpressionAttributeValues={
                ':room_id': room_id,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in',
                ':check_in': check_in,
                ':check_out': check_out
            }
        )
        
        return len(response['Items']) == 0
        
    except Exception as e:
        print(f"Availability check error: {str(e)}")
        return False

def is_admin(event):
    """Check if the user has admin role"""
    try:
        claims = event.get('requestContext', {}).get('authorizer', {}).get('claims', {})
        role = claims.get('custom:role', 'USER')
        return role == 'ADMIN'
    except:
        return False

def transform_room_for_frontend(room):
    """Transform room data to match exact frontend format from data/rooms.js"""
    return {
        'id': room.get('room_id'),
        'title': room.get('title'),
        'subtitle': room.get('subtitle', ''),
        'description': room.get('description'),
        'dogsAmount': room.get('dogsAmount', 1),
        'size': room.get('size'),
        'price': room.get('price', room.get('price_per_night')),  # Frontend expects 'price'
        'image': room.get('image'),  # S3 URL or base64
        'included': room.get('included', []),
        'reviews': room.get('reviews', [])
    }

def upload_image_to_s3(base64_image, key_prefix):
    """Upload base64 image to S3 and return the URL"""
    try:
        # Extract image data from base64 string
        header, encoded = base64_image.split(',', 1)
        image_data = base64.b64decode(encoded)
        
        # Determine file extension from header
        if 'jpeg' in header or 'jpg' in header:
            ext = 'jpg'
        elif 'png' in header:
            ext = 'png'
        else:
            ext = 'jpg'  # default
        
        # Create S3 key
        s3_key = f"{key_prefix}.{ext}"
        
        # Upload to S3
        s3.put_object(
            Bucket=S3_BUCKET,
            Key=s3_key,
            Body=image_data,
            ContentType=f"image/{ext}",
            ACL='public-read'
        )
        
        # Return S3 URL
        return f"https://{S3_BUCKET}.s3.amazonaws.com/{s3_key}"
        
    except Exception as e:
        print(f"Error uploading image to S3: {str(e)}")
        return base64_image  # Return original if upload fails

def convert_decimals(obj):
    """Convert Decimal objects to float for JSON serialization"""
    if isinstance(obj, list):
        return [convert_decimals(item) for item in obj]
    elif isinstance(obj, dict):
        return {key: convert_decimals(value) for key, value in obj.items()}
    elif isinstance(obj, Decimal):
        return float(obj)
    else:
        return obj

File: aws/src/handlers/rooms_handler.js
const { DynamoDBClient } = require('@aws-sdk/client-dynamodb');
const { DynamoDBDocumentClient, GetCommand, ScanCommand, PutCommand, UpdateCommand, DeleteCommand, QueryCommand } = require('@aws-sdk/lib-dynamodb');
const { S3Client, PutObjectCommand } = require('@aws-sdk/client-s3');
const { v4: uuidv4 } = require('uuid');
const { corsResponse, corsErrorResponse, handlePreflightRequest, extractOriginFromEvent, isAdmin } = require('./cors_utils');

// Initialize AWS clients
const ddbClient = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(ddbClient);
const s3Client = new S3Client({});

// Environment variables
const ROOMS_TABLE = process.env.DYNAMODB_TABLE_ROOMS || 'DogotelRooms';
const BOOKINGS_TABLE = process.env.DYNAMODB_TABLE_BOOKINGS || 'DogotelBookings';
const S3_BUCKET = process.env.S3_BUCKET || 'dogotel-images';

exports.handler = async (event, context) => {
    console.log('Event:', JSON.stringify(event, null, 2));

    try {
        const httpMethod = event.httpMethod;
        const path = event.path;
        const origin = extractOriginFromEvent(event);

        console.log(`Processing ${httpMethod} ${path} from origin: ${origin}`);

        // Handle preflight OPTIONS requests
        if (httpMethod === 'OPTIONS') {
            return handlePreflightRequest(event, origin);
        }

        if (path === '/rooms' && httpMethod === 'GET') {
            return await handleListRooms(event);
        } else if (path.startsWith('/rooms/') && path.split('/').length === 3 && httpMethod === 'GET') {
            const roomId = path.split('/')[2];
            return await handleGetRoom(roomId, event);
        } else if (path.startsWith('/rooms/') && path.endsWith('/unavailable-dates') && httpMethod === 'GET') {
            const roomId = path.split('/')[2];
            return await handleGetUnavailableDates(roomId, event);
        } else if (path.startsWith('/rooms/') && path.endsWith('/unavailable-ranges') && httpMethod === 'GET') {
            const roomId = path.split('/')[2];
            return await handleGetUnavailableRanges(roomId, event);
        } else if (path.startsWith('/rooms/') && path.endsWith('/reviews') && httpMethod === 'POST') {
            const roomId = path.split('/')[2];
            return await handleAddReview(roomId, event);
        } else if (path.startsWith('/rooms/') && path.endsWith('/reviews') && httpMethod === 'GET') {
            const roomId = path.split('/')[2];
            return await handleGetReviews(roomId, event);
        } else if (path === '/rooms' && httpMethod === 'POST') {
            return await handleCreateRoom(event);
        } else if (path.startsWith('/rooms/') && path.split('/').length === 3 && httpMethod === 'PUT') {
            const roomId = path.split('/')[2];
            return await handleUpdateRoom(roomId, event);
        } else if (path.startsWith('/rooms/') && path.split('/').length === 3 && httpMethod === 'DELETE') {
            const roomId = path.split('/')[2];
            return await handleDeleteRoom(roomId, event);
        } else {
            return corsErrorResponse(404, 'Endpoint not found', origin);
        }
    } catch (error) {
        console.error('Error in rooms handler:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
};

async function handleListRooms(event) {
    try {
        const origin = extractOriginFromEvent(event);
        const queryParams = event.queryStringParameters || {};
        const { min_price, max_price, check_in, check_out, size } = queryParams;

        // Get all rooms
        const result = await docClient.send(new ScanCommand({
            TableName: ROOMS_TABLE
        }));

        let rooms = result.Items || [];

        // Apply filters
        const filteredRooms = [];
        for (const room of rooms) {
            // Price filter
            const roomPrice = room.price || room.price_per_night || 0;
            if (min_price && roomPrice < parseFloat(min_price)) continue;
            if (max_price && roomPrice > parseFloat(max_price)) continue;

            // Size filter
            if (size && room.size?.toLowerCase() !== size.toLowerCase()) continue;

            // Availability filter
            if (check_in && check_out) {
                const available = await isRoomAvailable(room.room_id, check_in, check_out);
                if (!available) continue;
            }

            // Transform room for frontend
            const transformedRoom = transformRoomForFrontend(room);
            filteredRooms.push(transformedRoom);
        }

        return corsResponse(200, filteredRooms, origin);
    } catch (error) {
        console.error('List rooms error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleGetRoom(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);

        const result = await docClient.send(new GetCommand({
            TableName: ROOMS_TABLE,
            Key: { room_id: roomId }
        }));

        if (!result.Item) {
            return corsErrorResponse(404, 'Room not found', origin);
        }

        const room = transformRoomForFrontend(result.Item);
        return corsResponse(200, room, origin);
    } catch (error) {
        console.error('Get room error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleGetUnavailableDates(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);

        // Get all confirmed bookings for this room
        const result = await docClient.send(new QueryCommand({
            TableName: BOOKINGS_TABLE,
            IndexName: 'RoomBookingsIndex',
            KeyConditionExpression: 'room_id = :room_id',
            FilterExpression: '#status IN (:confirmed, :checked_in)',
            ExpressionAttributeNames: { '#status': 'status' },
            ExpressionAttributeValues: {
                ':room_id': roomId,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in'
            }
        }));

        const unavailableDates = [];
        for (const booking of result.Items || []) {
            // Handle both old and new date formats
            const bookingCheckIn = new Date(
                booking.check_in || booking.room?.startDate || booking.startDate
            );
            const bookingCheckOut = new Date(
                booking.check_out || booking.room?.endDate || booking.endDate
            );

            if (isNaN(bookingCheckIn.getTime()) || isNaN(bookingCheckOut.getTime())) {
                console.log('Invalid dates in booking:', booking);
                continue;
            }

            let currentDate = new Date(bookingCheckIn);
            const endDate = new Date(bookingCheckOut);

            while (currentDate < endDate) {
                unavailableDates.push(currentDate.toISOString().split('T')[0]);
                currentDate.setDate(currentDate.getDate() + 1);
            }
        }

        return corsResponse(200, {
            roomId: roomId,
            unavailableDates: [...new Set(unavailableDates)].sort()
        }, origin);
    } catch (error) {
        console.error('Get unavailable dates error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleGetUnavailableRanges(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);

        // Get all confirmed bookings for this room
        const result = await docClient.send(new QueryCommand({
            TableName: BOOKINGS_TABLE,
            IndexName: 'RoomBookingsIndex',
            KeyConditionExpression: 'room_id = :room_id',
            FilterExpression: '#status IN (:confirmed, :checked_in)',
            ExpressionAttributeNames: { '#status': 'status' },
            ExpressionAttributeValues: {
                ':room_id': roomId,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in'
            }
        }));

        const unavailableRanges = [];
        for (const booking of result.Items || []) {
            // Handle both old and new date formats
            const bookingCheckIn = new Date(
                booking.check_in || booking.room?.startDate || booking.startDate
            );
            const bookingCheckOut = new Date(
                booking.check_out || booking.room?.endDate || booking.endDate
            );

            if (isNaN(bookingCheckIn.getTime()) || isNaN(bookingCheckOut.getTime())) {
                console.log('Invalid dates in booking:', booking);
                continue;
            }

            // Adjust checkout date (booking ends day before checkout)
            const actualEndDate = new Date(bookingCheckOut);
            actualEndDate.setDate(actualEndDate.getDate() - 1);

            unavailableRanges.push({
                start: bookingCheckIn.toISOString().split('T')[0],
                end: actualEndDate.toISOString().split('T')[0]
            });
        }

        return corsResponse(200, {
            roomId: roomId,
            unavailableRanges: unavailableRanges
        }, origin);
    } catch (error) {
        console.error('Get unavailable ranges error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleCreateRoom(event) {
    try {
        const origin = extractOriginFromEvent(event);

        if (!isAdmin(event)) {
            return corsErrorResponse(403, 'Admin access required', origin);
        }

        const body = JSON.parse(event.body);
        const roomId = uuidv4();

        const room = {
            room_id: roomId,
            title: body.title,
            subtitle: body.subtitle,
            description: body.description,
            dogsAmount: body.dogsAmount,
            price: body.price,
            size: body.size,
            image: body.image,
            included: body.included || [],
            reviews: body.reviews || [],
            is_available: body.is_available !== false,
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString()
        };

        await docClient.send(new PutCommand({
            TableName: ROOMS_TABLE,
            Item: room
        }));

        return corsResponse(201, {
            success: true,
            roomId: roomId,
            message: 'Room created successfully'
        }, origin);
    } catch (error) {
        console.error('Create room error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleUpdateRoom(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);

        if (!isAdmin(event)) {
            return corsErrorResponse(403, 'Admin access required', origin);
        }

        const body = JSON.parse(event.body);

        const updateExpression = [];
        const expressionAttributeValues = {};
        const expressionAttributeNames = {};

        Object.entries(body).forEach(([key, value]) => {
            if (key !== 'room_id') {
                updateExpression.push(`#${key} = :${key}`);
                expressionAttributeNames[`#${key}`] = key;
                expressionAttributeValues[`:${key}`] = value;
            }
        });

        expressionAttributeNames['#updatedAt'] = 'updatedAt';
        expressionAttributeValues[':updatedAt'] = new Date().toISOString();
        updateExpression.push('#updatedAt = :updatedAt');

        await docClient.send(new UpdateCommand({
            TableName: ROOMS_TABLE,
            Key: { room_id: roomId },
            UpdateExpression: `SET ${updateExpression.join(', ')}`,
            ExpressionAttributeNames: expressionAttributeNames,
            ExpressionAttributeValues: expressionAttributeValues
        }));

        return corsResponse(200, {
            success: true,
            message: 'Room updated successfully'
        }, origin);
    } catch (error) {
        console.error('Update room error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleDeleteRoom(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);

        if (!isAdmin(event)) {
            return corsErrorResponse(403, 'Admin access required', origin);
        }

        await docClient.send(new DeleteCommand({
            TableName: ROOMS_TABLE,
            Key: { room_id: roomId }
        }));

        return corsResponse(200, {
            success: true,
            message: 'Room deleted successfully'
        }, origin);
    } catch (error) {
        console.error('Delete room error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function isRoomAvailable(roomId, checkIn, checkOut) {
    try {
        const result = await docClient.send(new QueryCommand({
            TableName: BOOKINGS_TABLE,
            IndexName: 'RoomBookingsIndex',
            KeyConditionExpression: 'room_id = :room_id',
            FilterExpression: '#status IN (:confirmed, :checked_in)',
            ExpressionAttributeNames: { '#status': 'status' },
            ExpressionAttributeValues: {
                ':room_id': roomId,
                ':confirmed': 'confirmed',
                ':checked_in': 'checked_in'
            }
        }));

        const requestCheckIn = new Date(checkIn);
        const requestCheckOut = new Date(checkOut);

        for (const booking of result.Items || []) {
            // Handle both old and new date formats
            const bookingCheckIn = new Date(
                booking.check_in || booking.room?.startDate || booking.startDate
            );
            const bookingCheckOut = new Date(
                booking.check_out || booking.room?.endDate || booking.endDate
            );

            if (isNaN(bookingCheckIn.getTime()) || isNaN(bookingCheckOut.getTime())) {
                console.log('Invalid dates in booking:', booking);
                continue;
            }

            // Check for overlap
            if (requestCheckIn < bookingCheckOut && requestCheckOut > bookingCheckIn) {
                return false;
            }
        }

        return true;
    } catch (error) {
        console.error('Room availability check error:', error);
        return false;
    }
}

function transformRoomForFrontend(room) {
    return {
        id: room.room_id,
        title: room.title,
        subtitle: room.subtitle,
        description: room.description,
        dogsAmount: room.dogsAmount,
        price: room.price,
        size: room.size,
        image: room.image,
        included: room.included || [],
        reviews: room.reviews || [],
        isAvailable: room.is_available !== false
    };
}

async function handleAddReview(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);
        const body = JSON.parse(event.body);
        const { email, stars, text } = body;

        if (!email || !stars || !text) {
            return corsErrorResponse(400, 'Missing required fields: email, stars, text', origin);
        }

        if (stars < 1 || stars > 5) {
            return corsErrorResponse(400, 'Stars must be between 1 and 5', origin);
        }

        // Check if room exists
        const roomResult = await docClient.send(new GetCommand({
            TableName: ROOMS_TABLE,
            Key: { room_id: roomId }
        }));

        if (!roomResult.Item) {
            return corsErrorResponse(404, 'Room not found', origin);
        }

        // Create new review
        const newReview = {
            name: email.split('@')[0], // Use email prefix as name
            email: email,
            stars: parseInt(stars),
            review: text,
            date: new Date().toISOString(),
            reviewId: uuidv4()
        };

        // Get current reviews and add new one
        const currentReviews = roomResult.Item.reviews || [];
        
        // Check if user already reviewed this room
        const existingReviewIndex = currentReviews.findIndex(r => r.email === email);
        if (existingReviewIndex !== -1) {
            // Update existing review
            currentReviews[existingReviewIndex] = newReview;
        } else {
            // Add new review
            currentReviews.push(newReview);
        }

        // Update room with new reviews
        await docClient.send(new UpdateCommand({
            TableName: ROOMS_TABLE,
            Key: { room_id: roomId },
            UpdateExpression: 'SET reviews = :reviews, updatedAt = :updatedAt',
            ExpressionAttributeValues: {
                ':reviews': currentReviews,
                ':updatedAt': new Date().toISOString()
            }
        }));

        return corsResponse(201, {
            success: true,
            message: existingReviewIndex !== -1 ? 'Review updated successfully' : 'Review added successfully',
            review: newReview
        }, origin);

    } catch (error) {
        console.error('Add review error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}

async function handleGetReviews(roomId, event) {
    try {
        const origin = extractOriginFromEvent(event);

        // Get room with reviews
        const result = await docClient.send(new GetCommand({
            TableName: ROOMS_TABLE,
            Key: { room_id: roomId }
        }));

        if (!result.Item) {
            return corsErrorResponse(404, 'Room not found', origin);
        }

        const reviews = result.Item.reviews || [];

        return corsResponse(200, {
            success: true,
            roomId: roomId,
            reviews: reviews,
            reviewCount: reviews.length,
            averageRating: reviews.length > 0 
                ? (reviews.reduce((sum, r) => sum + r.stars, 0) / reviews.length).toFixed(1)
                : 0
        }, origin);

    } catch (error) {
        console.error('Get reviews error:', error);
        return corsErrorResponse(500, 'Internal server error', extractOriginFromEvent(event));
    }
}